From 48a42f10b644cb49a687db2bcad08c8170043014 Mon Sep 17 00:00:00 2001
From: Jiafei Pan <Jiafei.Pan@nxp.com>
Date: Thu, 12 Mar 2026 17:58:59 +0800
Subject: [PATCH 03/10] feat(imx93): add common interrupt handler and enable it
 on imx93

Added a common interrupt handler for all EL3 interrupt, so it can handle
more EL3 interrupts.

Enable it on i.MX93.

Signed-off-by: Jiafei Pan <Jiafei.Pan@nxp.com>
Upstream-Status: Pending
---
 plat/imx/common/imx_interrupt_mgmt.c         | 72 ++++++++++++++++++++
 plat/imx/common/include/imx_interrupt_mgmt.h | 23 +++++++
 plat/imx/imx93/imx93_bl31_setup.c            |  5 +-
 plat/imx/imx93/platform.mk                   |  1 +
 4 files changed, 100 insertions(+), 1 deletion(-)
 create mode 100644 plat/imx/common/imx_interrupt_mgmt.c
 create mode 100644 plat/imx/common/include/imx_interrupt_mgmt.h

diff --git a/plat/imx/common/imx_interrupt_mgmt.c b/plat/imx/common/imx_interrupt_mgmt.c
new file mode 100644
index 000000000..33c1a837f
--- /dev/null
+++ b/plat/imx/common/imx_interrupt_mgmt.c
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2024-2026 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+#include <arch_helpers.h>
+
+#include <bl31/interrupt_mgmt.h>
+#include <common/debug.h>
+#include <imx_interrupt_mgmt.h>
+#include <plat/common/platform.h>
+
+static interrupt_type_handler_t type_el3_interrupt_table[MAX_INTR_EL3];
+
+int request_intr_type_el3(uint32_t id, interrupt_type_handler_t handler)
+{
+	/* Validate 'handler' and 'id' parameters */
+	if (!handler || id >= MAX_INTR_EL3) {
+		return -EINVAL;
+	}
+
+	/* Check if a handler has already been registered */
+	if (type_el3_interrupt_table[id] != NULL) {
+		return -EALREADY;
+	}
+
+	type_el3_interrupt_table[id] = handler;
+
+	return 0;
+}
+
+static uint64_t imx_el3_interrupt_handler(uint32_t id, uint32_t flags,
+						void *handle, void *cookie)
+{
+	uint32_t intr_id;
+	interrupt_type_handler_t handler;
+
+	intr_id = plat_ic_get_pending_interrupt_id();
+	if (intr_id < 0) {
+		WARN("No pending interrupt\n");
+		return 0U;
+	}
+	intr_id = plat_ic_acknowledge_interrupt();
+
+	INFO("Interrupt recvd is %d\n", intr_id);
+
+	handler = type_el3_interrupt_table[intr_id];
+	if (handler != NULL) {
+		handler(intr_id, flags, handle, cookie);
+	}
+
+	/*
+	 * Mark this interrupt as complete to avoid a interrupt storm.
+	 */
+	plat_ic_end_of_interrupt(intr_id);
+
+	return 0U;
+}
+
+void imx_el3_interrupt_config(void)
+{
+	uint64_t flags = 0U;
+	uint64_t rc;
+
+	set_interrupt_rm_flag(flags, NON_SECURE);
+	rc = register_interrupt_type_handler(INTR_TYPE_EL3,
+						imx_el3_interrupt_handler, flags);
+	if (rc != 0U) {
+		panic();
+	}
+}
diff --git a/plat/imx/common/include/imx_interrupt_mgmt.h b/plat/imx/common/include/imx_interrupt_mgmt.h
new file mode 100644
index 000000000..4799db5fc
--- /dev/null
+++ b/plat/imx/common/include/imx_interrupt_mgmt.h
@@ -0,0 +1,23 @@
+/*
+ * Copyright 2024-2026 NXP
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+#ifndef IMX_EL3_INTRPT_MGMT_H
+#define IMX_EL3_INTRPT_MGMT_H
+
+#include <bl31/interrupt_mgmt.h>
+
+#define MAX_INTR_EL3		128
+
+/*
+ * Register handler to specific GIC entrance
+ * for INTR_TYPE_EL3 type of interrupt
+ */
+int request_intr_type_el3(uint32_t id, interrupt_type_handler_t handler);
+
+void imx_el3_interrupt_config(void);
+
+#endif	/*	IMX_EL3_INTRPT_MGMT_H	*/
diff --git a/plat/imx/imx93/imx93_bl31_setup.c b/plat/imx/imx93/imx93_bl31_setup.c
index c134512f2..954d7b608 100644
--- a/plat/imx/imx93/imx93_bl31_setup.c
+++ b/plat/imx/imx93/imx93_bl31_setup.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022-2023 NXP
+ * Copyright 2022-2024 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -21,6 +21,7 @@
 
 #include <ele_api.h>
 #include <dram.h>
+#include <imx_interrupt_mgmt.h>
 #include <imx8_lpuart.h>
 #include <plat_common.h>
 #include <plat_imx8.h>
@@ -157,6 +158,8 @@ void bl31_platform_setup(void)
 	/* get soc info */
 	ele_get_soc_info();
 
+	imx_el3_interrupt_config();
+
 	/* Init the dram info */
 	dram_info_init(SAVED_DRAM_TIMING_BASE);
 
diff --git a/plat/imx/imx93/platform.mk b/plat/imx/imx93/platform.mk
index 9db734be4..8c9a7bbd9 100644
--- a/plat/imx/imx93/platform.mk
+++ b/plat/imx/imx93/platform.mk
@@ -31,6 +31,7 @@ BL31_SOURCES		+=	common/desc_image_load.c			\
 				plat/imx/imx93/aarch64/plat_helpers.S		\
 				plat/imx/imx93/plat_topology.c			\
 				plat/imx/common/imx_common.c			\
+				plat/imx/common/imx_interrupt_mgmt.c		\
 				plat/imx/common/lpuart_console.S		\
 				plat/imx/imx93/trdc.c			\
 				plat/imx/imx93/pwr_ctrl.c			\
-- 
2.43.0

